src/distributions/uniform.rs

```diff
@@ -737,52 +742,75 @@ impl UniformSampler for UniformDuration {
         let low = *low_b.borrow();
         let high = *high_b.borrow();
         assert!(low <= high, "Uniform::new_inclusive called with `low > high`");
-        let size = high - low;
-        let nanos = size
-            .as_secs()
-            .checked_mul(1_000_000_000)
-            .and_then(|n| n.checked_add(size.subsec_nanos() as u64));
-
-        let mode = match nanos {
-            Some(nanos) => {
-                UniformDurationMode::Small {
-                    nanos: Uniform::new_inclusive(0, nanos),
-                }
+
+        let low_s = low.as_secs();
+        let low_n = low.subsec_nanos();
+        let mut high_s = high.as_secs();
+        let mut high_n = high.subsec_nanos();
+
+        if high_n < low_n {
+            high_s = high_s - 1;
+            high_n = high_n + 1_000_000_000;
+        } else {
+            high_s = high_s;
+            high_n = high_n;
+        }
+
+        let mode = if low_s == high_s {
+            UniformDurationMode::Small {
+                secs: low_s,
+                nanos: Uniform::new_inclusive(low_n, high_n),
             }
-            None => {
+        } else {
+            let max = high_s
+                .checked_mul(1_000_000_000)
+                .and_then(|n| n.checked_add(high_n as u64));
+
+            if let Some(higher_bound) = max {
+                let lower_bound = low_s * 1_000_000_000 + low_n as u64;
+                UniformDurationMode::Medium {
+                    nanos: Uniform::new_inclusive(lower_bound, higher_bound),
+                }
+            } else {
+                // An offset is applied to simplify generation of nanoseconds
+                let max_nanos = high_n - low_n;
                 UniformDurationMode::Large {
-                    size: size,
-                    secs: Uniform::new_inclusive(0, size.as_secs()),
+                    max_secs: high_s,
+                    max_nanos,
+                    secs: Uniform::new_inclusive(low_s, high_s),
                 }
             }
         };
-
         UniformDuration {
             mode,
-            offset: low,
+            offset: low_n,
         }
     }
 
     #[inline]
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Duration {
-        let d = match self.mode {
-            UniformDurationMode::Small { nanos } => {
+        match self.mode {
+            UniformDurationMode::Small { secs, nanos } => {
+                let n = nanos.sample(rng);
+                Duration::new(secs, n)
```

Comment:
Unfortunately this is now incorrect since it's possible that n > 10^9 (in this case `new` should panic). Hopefully using an `if` to check here won't have too big an impact.

Reply:
`Duration::new()` actually handles that already. In the documentation it says:
> If the number of nanoseconds is greater than 1 billion (the number of nanoseconds in a second), then it will carry over into the seconds provided.

Reply:
Maybe add a `debug_assert`?